src/lib.rs

```diff
@@ -934,4 +867,68 @@ mod tests {
 
         system.shutdown().unwrap();
     }
+
+    #[test]
+    fn timeouts() {
+        struct TimeoutActor {
+            handle_count: Arc<AtomicU32>,
+            timeout_count: Arc<AtomicU32>,
+        }
+
+        impl Actor for TimeoutActor {
+            type Error = ();
+            type Message = Option<Instant>;
+
+            fn name() -> &'static str {
+                "TimeoutActor"
+            }
+
+            fn handle(&mut self, ctx: &mut Context<Self>, msg: Option<Instant>) -> Result<(), ()> {
+                self.handle_count.fetch_add(1, Ordering::SeqCst);
+                if msg.is_some() {
+                    ctx.receive_deadline = msg;
+                }
+                Ok(())
+            }
+
+            fn deadline_passed(&mut self, _: &mut Context<Self>, _: Instant) {
+                self.timeout_count.fetch_add(1, Ordering::SeqCst);
+            }
+        }
+
+        let mut system = System::new("timeouts");
+        let (handle_count, timeout_count) = (Default::default(), Default::default());
+        let actor = TimeoutActor {
+            handle_count: Arc::clone(&handle_count),
+            timeout_count: Arc::clone(&timeout_count),
+        };
+        let addr = system.spawn(actor).unwrap();
+
+        // Test that setting deadline to past triggers the deadline immediately.
+        addr.send(Some(Instant::now() - Duration::from_secs(1))).unwrap();
+        thread::sleep(Duration::from_millis(10));
+        assert_eq!(handle_count.load(Ordering::SeqCst), 1);
+        assert_eq!(timeout_count.load(Ordering::SeqCst), 1);
+
+        // Test the normal case.
+        addr.send(Some(Instant::now() + Duration::from_millis(20))).unwrap();
+        thread::sleep(Duration::from_millis(10));
+        assert_eq!(handle_count.load(Ordering::SeqCst), 2);
+        assert_eq!(timeout_count.load(Ordering::SeqCst), 1);
```

Comment:
This test can be flaky. Does Rust testing framework support marking such tests?

Reply:
> This test can be flaky.

The intention is that they are not flaky in practice - that we sleep at least 10 ms between writing a value in one thread and asserting on it in another thread. I didn't hit a failed test even if I lowered the delay to 1 ms (with 0 ms wait I did). That isn't of course a proof that it won't be flaky.

Perhaps I've overlooked something? My naive assumption was that a failure can occur only when OS scheduler is very unfair and schedules thread A _after_ thread B that requested to be woken 10 ms before A requested to be woken.

> Does Rust testing framework support marking such tests?

I have only found an `#[ignore]` attribute, which is rather coarse https://doc.rust-lang.org/reference/attributes/testing.html#the-ignore-attribute
More loosely related, I had nice experience with parametrized testing using https://github.com/frondeus/test-case

Reply:
You didn't overlook anything. I was referring to the case when the OS scheduler is very unfair e.g. <1% of time on a significantly overloaded CI machine :) Thanks for testing with 1ms granularity and also checking the test passes. It's good as is, and this conversation is a useful reference for later :)